_posts/2020-03-29-all-you-need-is-lambda-1-booleans.md

```diff
@@ -0,0 +1,260 @@
+
+```
+Bool = ∀ a . a -> a -> a
+```
+
+> Aside: I’m making explicit use of the for-all quantifier here to drive home the point that any particular `Bool` value must be able to be applied to `then` and `else` values of any arbitrary type `a`, defined now or in the future.
+>
+> By the same token, we could have written `if`’s type more explicitly as:
+>
+> ```
+> if : ∀ a . Bool -> a -> a -> a
+> ```
+>
+> Here and in future, local type variables can be assumed to be implicitly generalized in the same manner as Haskell if not otherwise quantified.
+
+If a given `Bool` is a function of two arguments of arbitrary type, returning the same type, it must therefore select one of its arguments to return. There are only two distinguishable inhabitants of `Bool`, `true` and `false`, so we can therefore deduce that since `if` defers the selection of the result to the `Bool`, for `true` and `false` to actually differ they must make opposite selections. In other words, `true` must return the `then` parameter, while `false` must return the `else` one:
+
+
+```
+true, false : Bool
+true  = λ then else . then
+false = λ then else . else
+```
+
+We didn’t move the problem around at all—we actually solved it (once we defined `true` and `false`, anyway). It’s simply that encoding booleans in this manner makes `if` largely pointless, since if we can apply `if` to a `Bool` and two arguments, we could equally apply the `Bool` to those arguments directly.
+
+> Aside: we chose to define `if` as applying the `Bool` to the other arguments in the same order it received them, but we could just as easily have swapped them:
+>
+> ```
+> if = λ cond then else . cond else then
+> ```
+>
+> In this case, `if` would be more useful since it would preserve our familiar argument ordering. As an exercise for the reader, consider what other effects this difference would have. What are the tradeoffs, syntactically and semantically? When would one or the other definition be more or less convenient?
+
+It’s frequently convenient to conflate booleans with bits, their minimal representation, but in truth they’re not the same at all. Practically, some programming languages define booleans as a byte in memory, perhaps clamping its values to 0 and 1; others define them as instances of some boolean class, or constructors of an algebraic datatype. Some provide no formal relationship between `true` and `false` at all, save for a common interface—duck typing.
+
+Mathematically, booleans are the values in propositional logic; the upper and lower bounds of a lattice; the zero and one of a semiring; the members of the set with cardinality 2; and many other things in many different contexts.
+
```

Comment:
I would emphasize `representing it with functions which can perform all the operations possible with the original datatype`, since it's the linchpin of this piece.